node/src/components/small_network/counting_format.rs

```diff
@@ -87,26 +128,179 @@ where
+        // something in our assessment that hashing these should be safe. Additionally, these values
+        // are XOR'd, not concatenated. All this is done to prevent leaking information about these
+        // numbers.
+        //
+        // Some SSL implementations use timestamps for the first four bytes, so to be sufficiently
+        // random, we use 4 + 8 bytes of the nonces.
+        let mut server_random = [0; 12];
+        let mut client_random = [0; 12];
+
+        ssl.server_random(&mut server_random);
+        ssl.client_random(&mut client_random);
+
+        // Hash the resulting random values.
+        let mut id = hash::hash(server_random).to_array();
+
+        // We XOR in a hashes of server and client fingerprint, to ensure that in the case of an
+        // accidental collision (e.g. when `server_random` and `client_random` turn out to be all
+        // zeros), we still have a chance of producing a reasonable ID.
+        if let Some(our_id_bytes) = our_id.hash_bytes() {
+            utils::xor(&mut id, &our_id_bytes[0..Digest::LENGTH]);
+        } else {
+            error!(
+                ?our_id,
+                "small_network attempted to retrieve bytes of ID, but seems to be libp2p ID?"
+            )
+        }
+
+        if let Some(their_id_bytes) = their_id.hash_bytes() {
+            utils::xor(&mut id, &their_id_bytes[0..Digest::LENGTH]);
+        } else {
+            error!(
+                ?their_id,
+                "small_network attempted to retrieve bytes of ID, but seems to be libp2p ID?"
+            )
+        }
+
+        ConnectionId(id)
+    }
+
+    /// Creates a new [`TraceID`] based on the message count.
+    fn create_trace_id(&self, flag: u8, count: u64) -> TraceId {
+        // Copy the basic network ID.
+        let mut buffer = self.0;
+
+        // Direction set on first byte.
+        buffer[0] ^= flag;
+
+        // XOR in message count.
+        utils::xor(&mut buffer[4..12], &count.to_ne_bytes());
+
+        // Hash again and truncate.
+        let full_hash = hash::hash(&buffer);
+
+        // Safe to expect here, as we assert earlier that `Digest` is at least 12 bytes.
+        let truncated =
+            TryFrom::try_from(&full_hash.to_array()[0..8]).expect("buffer size mismatch");
+
+        TraceId(truncated)
+    }
+
+    /// Creates a random `ConnectionId`.
+    #[cfg(test)]
+    fn random(rng: &mut TestRng) -> Self {
+        let rand_digest = Digest::random(rng);
+        Self(rand_digest.to_array())
+    }
+}
+
+/// Message sending direction.
+#[derive(Copy, Clone, Debug)]
+#[repr(u8)]
+pub(super) enum Role {
+    /// Dialer, i.e. initiator of the connection.
+    Dialer,
+    /// Listener, acceptor of the connection.
+    Listener,
+}
+
+impl Role {
+    #[inline]
+    fn in_flag(self) -> u8 {
+        !(self.out_flag())
+    }
+
+    #[inline]
+    fn out_flag(self) -> u8 {
+        match self {
+            Role::Dialer => 0xaa,
+            Role::Listener => !0xaa,
```

Comment:
Why this complicated? Why not simply `0u8` and `1u8`?

Reply:
Having them be the inverse of each other is a useful property for the code below, so it'd be `0` and `256` instead. I thought it'd be neat to have the `1010_1010` and `0101_0101` bit patterns here.

Reply:
Oh, I missed that completely. Can you use the `0b` format then and/or add a comment?

Reply:
I'll do both!